src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1666,52 +1666,101 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
+                }
+                else
+                {
+                    // This code block adds FK definitions between source and target entities when there is no FK constraint defined
+                    // in the database, either from source->target or target->source entities.
+
+                    // Being here indicates that we did not find an FK constraint in the database for the current FK definition.
+                    // But this does not indicate absence of an FK constraint between the source, target entities yet.
+                    // This may happen when an FK constraint exists between two tables, but in an order opposite to the order
+                    // of referencing and referenced tables present in the current FK definition. This happens because for a relationship
+                    // with right cardinality as 1, we add FK definitons from both source->target and target->source to the source entity's definition.
+                    // because at that point we don't know if the relationship is an N:1 relationship or a 1:1 relationship.
+                    // So here, we need to remove the wrong FK definition for:
+                    // 1. N:1 relationships,
+                    // 2. 1:1 relationships where an FK constraint exists only from source->target or target->source but not both.
+
+                    // E.g. for a relationship between Book-Publisher entities with cardinality 1, we would have added a Foreign key definition
+                    // from Book->Publisher and Publisher->Book to Book's source definition earlier.
+                    // Since it is an N:1 relationship, it might have been the case that the current FK definition had
+                    // 'publishers' table as the referencing table and 'books' table as the referenced table, and hence,
+                    // we did not find any FK constraint. But an FK constraint does exist where 'books' is the referencing table
+                    // while the 'publishers' is the referenced table.
+                    // (The definition for that constraint would be taken care of while adding database FKs above.)
+
+                    // So, before concluding that there is no FK constraint between the source, target entities, we need
+                    // to confirm absence of FK constraint from source->target and target->source tables.
+                    RelationShipPair inverseFKPair = new(fKDefinitionToTarget.Pair.ReferencedDbTable, fKDefinitionToTarget.Pair.ReferencingDbTable);
+
+                    // Add FK definition to the set of validated FKs only if no FK constraint is defined for the source and target entities
+                    // in the database, either from source -> target or target -> source.
+                    if (PairToFkDefinition is not null && !PairToFkDefinition.ContainsKey(inverseFKPair))
+                    {
+                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
```

Comment:
If there is no FK definition either from source-> target or target->source, adding the fkDefinition here will mean its coming from config. But if it doesn't contain any referencing/referenced columns, adding it here is of no use. instead that should be a validation error.

Reply:
Like we discussed, this should be caught during config validation. Until we have that validation checked in (issue: https://github.com/Azure/data-api-builder/issues/1935), we can decide to have a validation in place here. What do you suggest? I'll ping Sean to see if this can be assigned to anyone for 0.12 in the meantime.

Reply:
So pending #1935 , I understand that certain invalid state should be caught during validation such that it isn't needed to be dealt with here:
1. as Ani mentioned, when no db fk exists, but relationship is defined in config.. and that relationship doesn't define source or target fields. OR source.fields.count != target.fields.count as mentioned by the linked validation issue.

Ayush, this is also a condition that is not hit in the integration tests. Can you add a test to exercise this condition? e.g. no db defined fk for N:1 relationship

Reply:
As I mentioned in the PR, I plan to add tests for custom relationships once this PR: https://github.com/Azure/data-api-builder/pull/2056 is in. Currently, there are no set entities present in the config which allow me to add meaningful custom relationships. Hence, I have manually tested the scenarios. 

Also for custom relationships, 1:1 or N:1 does not really matter as I explained during the session. Let me know if you want to get on a call and run through some manual tests which I did.